### DISCUSSION TITLE
Chance of using github releases or s3

### DISCUSSION THREAD

**Discussion by KelvinSan** [dennisvang/tufup#115] [Q&A]
Is there any chance that this will support github releases, S3 buckets and release channels like pyupdater. Also, I know this sounds like an odd question but what are the chances this repo gets archived like pyupdater I would hate to have to find a new way to update python apps haha. Thanks for your time in advance

**Comment by dennisvang** [comment]
> Is there any chance that this will support github releases, S3 buckets and release channels like pyupdater. Also, I know this sounds like an odd question but what are the chances this repo gets archived like pyupdater I would hate to have to find a new way to update python apps haha. Thanks for your time in advance

Hi @KelvinSan,

In order to keep the maintenance burden to a minimum, we aim to keep tufup lean and clean. Therefore we do not plan to include built-in support for 3rd party services like github-releases or AWS S3 buckets. 

Nevertheless, tufup can be used with these services. We use it ourselves in combination with S3.

The reason for minimizing the maintenance burden is so we can keep tufup up and running for as long as possible. As we depend on tufup ourselves, we have a good reason to keep it up-and-running. Of course we also hope to get some more contributors to help with this.

> ... and release channels like pyupdater.

Like PyUpdater, `tufup` also supports "release channels," but we use [PEP440 terminology][2]. By default, the update client will search the "(final) release" channel (pyupdater's "stable"). You have the option to specify a "pre-release" channel ("a", "b", "rc") when checking for updates. Refer to the [`Client.check_for_updates()'][1] method for details:

> Final releases are always included. Pre-releases are excluded by default. If `pre` is specified, pre-releases are included, ***down to the specified level***. Pre-release identifiers follow the PEP440 specification, i.e. 'a', 'b', or 'rc', for alpha, beta, and release        candidate, respectively.

For example, suppose your latest release is 1.1.0, and your latest pre-release is 2.0.0a3, and an app in the field has version 1.0.0. If this app checks either the default channel, the release-candidate ("rc") channel, or the beta ("b") channel, it finds 1.1.0 available. If the app checks the alpha channel ("a"), it finds 2.0.0a3.

Also have a look at the corresponding [tests][3] to see some example calls.

Note, by the way, that PyUpdater's release channels are broken. See e.g. https://github.com/Digital-Sapphire/PyUpdater/issues/315 and https://github.com/Digital-Sapphire/PyUpdater/issues/255. This was one of the reasons for creating tufup.

[1]: https://github.com/dennisvang/tufup/blob/master/src/tufup/client.py#L138 
[2]: https://peps.python.org/pep-0440/#public-version-identifiers
[3]: https://github.com/dennisvang/tufup/blob/4204d68cd33961dce0bf5650db09413845ca9cab/tests/test_client.py#L133

**Comment by dennisvang** [comment]
Also see this discussion: https://github.com/dennisvang/tufup-example/issues/3
